apps/framework-cli/src/infrastructure/olap/clickhouse/inserter.rs

```diff
@@ -1,76 +1,308 @@
-use crate::infrastructure::olap::clickhouse::client::ClickHouseClient;
+use crate::infrastructure::olap::clickhouse::client::ClickHouseClientTrait;
 use crate::infrastructure::olap::clickhouse::model::ClickHouseRecord;
+use std::collections::{HashMap, VecDeque};
 use std::time::Duration;
 
 use log::{debug, error};
+use rdkafka::error::KafkaError;
 use std::sync::Arc;
 use tokio::sync::Mutex;
 use tokio::time;
 
-use super::config::ClickHouseConfig;
+#[derive(Default)]
+pub struct Batch {
+    pub records: Vec<ClickHouseRecord>,
+    pub partition_offsets: HashMap<i32, i64>,
+}
 
-const MAX_FLUSH_INTERVAL_SECONDS: u64 = 1;
-const MAX_BATCH_SIZE: usize = 100000;
+impl Batch {
+    fn update_offset(&mut self, partition: i32, offset: i64) {
+        self.partition_offsets
+            .entry(partition)
+            .and_modify(|e| *e = (*e).max(offset))
+            .or_insert(offset);
+    }
+}
 
-pub type BatchRecords = Arc<Mutex<Vec<ClickHouseRecord>>>;
+pub type OffsetCommitCallback = Box<dyn Fn(i32, i64) -> Result<(), KafkaError> + Send + Sync>;
+pub type BatchQueue = Arc<Mutex<VecDeque<Batch>>>;
 
-pub struct Inserter {
-    buffer: BatchRecords,
+pub struct Inserter<C: ClickHouseClientTrait + 'static> {
+    queue: BatchQueue,
+    client: Arc<C>,
```

Comment:
we can probably save an `Arc` by pointing to the `Inserter` rather than the `Inserter` having two pointers.

Reply:
Feels like that's ok as is, to do what you are suggesting we would need to pull the `flush` function outside of the `Inserter`. We could iterate there if we see this as an issue.